lib/internal/wrap_js_stream.js

```diff
@@ -0,0 +1,220 @@
+'use strict';
+
+const assert = require('assert');
+const util = require('util');
+const Socket = require('net').Socket;
+const JSStream = process.binding('js_stream').JSStream;
+const uv = process.binding('uv');
+const debug = util.debuglog('stream_wrap');
+const errors = require('internal/errors');
+
+/* This class serves as a wrapper for when the C++ side of Node wants access
+ * to a standard JS stream. For example, TLS or HTTP do not operate on network
+ * resource conceptually, although that is the common case and what we are
+ * optimizing for; in theory, they are completely composable and can work with
+ * any stream resource they see.
+ *
+ * For the common case, i.e. a TLS socket wrapping around a net.Socket, we
+ * can skip going through the JS layer and let TLS access the raw C++ handle
+ * of a net.Socket. The flipside of this is that, to maintain composability,
+ * a way to create "fake" net.Socket instances that call back into a
```

Comment:
Nit: change to active voice: 
The flipside of this is that, to maintain composability, we need a way to create ....